impeller/display_list/display_list_dispatcher.cc

```diff
@@ -290,6 +290,37 @@ static std::vector<Matrix> ToRSXForms(const SkRSXform xform[], int count) {
   return result;
 }
 
+// Convert display list colors + stops into impeller colors and stops, taking
+// care to ensure that the stops always start with 0.0 and end with 1.0. If
+// the color stops were generated, only pass through the colors. This signals
+// that the colors are intended to be evenly space.d
+template <typename T>
+static void ConvertStops(T* gradient,
+                         std::vector<Color>* colors,
+                         std::vector<float>* stops) {
+  auto* dl_colors = gradient->colors();
+  if (gradient->generated_stops()) {
+    for (auto i = 0; i < gradient->stop_count(); i++) {
+      colors->emplace_back(ToColor(dl_colors[i]));
+    }
+    return;
+  }
+
+  auto* dl_stops = gradient->stops();
+  if (dl_stops[0] != 0.0) {
```

Comment:
A silly question: Shall we add some assertions that stops are non-empty? Otherwise array access is out of bound.

I have checked who calls this code but seems cannot figure out where such assertion is ensured. On the contrary, the old code (`for (auto i = 0; i < radialGradient->stop_count(); i++)`) indeed works well for zero stops.

Reply:
we're mostly protected by checks at the dart:ui boundary here: https://github.com/flutter/engine/blob/0f9fc7aff92fcf975a4ab8c9c41673ab64d048df/lib/ui/painting.dart#L4073

Still, you're right that it would be good to add some more asserts and checks

Reply:
Thanks! But the link you provide does not check that, stops is *non-null empty*, i.e. `[]`?

Reply:
I believe that non-null empty and null are treated identically. If there are no stops, assume evenly spaced stops.

Reply:
Well, my worry is:

* If `stop_count == 0`, the C++ code will have memory problem.
* painting.dart does check `colorStops==null`
* But painting.dart does not check `colorStops != null && colorStops.isEmpty` IMHO. For example, a user can call `Gradient.linear(..., colorStops: [])`, and all Dart-level checks will pass